### PR TITLE
Add lwm2m deviceshifu design

### PR DESCRIPTION
**Pull request by rhoninl**
**What this PR does / why we need it**:
lwM2M deviceShifu design doc
**Will this PR make the community happier**? 
yes
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**How is this PR tested**
- [ ] unit test
- [ ] e2e test
- [x] other (please specify)
documentation no need test
**Special notes for your reviewer**:

**Release note**:

```release-note
- add lwM2M deviceShifu design doc
```

### CODE REVIEW

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+- Support lwM2M v1.1.x or later version.
+- Datagram Transport Layer Security (DTLS) support.
+- Over TCP or other protocol.
+- Bootstrap Server.
+
+## Detail Design
+
+The `deviceShifu-lwM2M` will represent each lwM2M object as an instruction that can be accessed through a RESTful API. The API will use the POST method for write operations and the GET method for read operations. The supported data formats for lwM2M v1.0 are TLV, JSON, Plain text, and Opaque.
+
+deviceShifu-lwM2M's lwM2M server will handle Register, Update, De-register, Read, Write, Observe and Notify.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:27]: you mean external LwM2M server? Like mosquitto broker for MQTT. Not part of deviceShifu

**Review by tomqin93 — COMMENTED** [review]
@rhoninl added some comments, please include an architecture diagram too

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md (hunk 1 above)
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:27]: No it is a part of deviceShifu, for devices are LwM2M client and deviceShifu is server which device can connect to us

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+deviceShifu-lwM2M support two kind of mode normal and observe mode.
+- normal mode: just like the other deviceshifu, when call the instruction, deviceShifu will send Read or Write Request to deviceShifu and return response
+- observe mode: this mode will set the min and max notify time, then device will notify data when data changed or timeout. deviceShifu will record data, and return the data store in the memory cache when call the instruction. and this kind of mode also support read and write operation.
```
[the remark this answers is not in the document]
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:31]: no need to record latest data, this mode will use pull method to read data from device, when call the instruction we will send read command to get the latest data.

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+deviceShifu will host a lwM2M server and listen on udp 3683(lwM2M default port) and http server(deviceshifu) on 8080.
```
[the remark this answers is not in the document]
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:33]: yes deviceshifu-LwM2M will have both http server and LwM2M server as well

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    ds ->> d: Created
+
+    alt observe mode
+    ds ->> d: Observe[Write the Notification Attributes Settings]
+    d ->> ds: success
+    ds ->> d: Observe
+    loop Device Data Change or Timeout
+        d ->> ds: Notify
+        ds ->> d: Update Device Data
+    end
```
[the remark this answers is not in the document]
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:60]: write just like the normal mode to send a write request to device. for observe mode just provide a way to receive pushed data

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    EnableObserve       bool        `json:"EnableObserve,omitempty"` // optional enable observe mode default false
+    ObserveMinPeriod    int         `json:"ObserveMinPeriod,omitempty"` // optional work when enable observe default 10 seconds
+    ObserveMaxPeriod    int         `json:"ObserveMaxPeriod,omitempty"` // optional work when enable observe default 60 seconds
+}
+```
+
+### Serving requests
+
+deviceShifu-LwM2M would take RESTful-style requests just as other deviceShifu do.
+LwM2M supports both `GET` and `PUT` requests.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:121]: newline

**Review by tomqin93 — COMMENTED** [review]
@rhoninl added some comments

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+- Create a deviceshifu-lwm2m type to enable users to connect devices using the lwM2M protocol.
+- LwM2M protocol support both `read` and `write` requests.
+- Resource observation or Notification.
+- [LwM2M protocol using v1.0.x version](https://www.openmobilealliance.org/release/LightweightM2M/V1_0-20170208-A/OMA-TS-LightweightM2M-V1_0-20170208-A.pdf).
+- LwM2M protocol under UDP.
+  
+### Design Non-Goal
+
+- Support LwM2M v1.1.x or later version.
+- Datagram Transport Layer Security (DTLS) support.
```
[the remark this answers is not in the document]
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:20]: yes we support DLTS with PSK

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+- Bootstrap Server.
+
+## Detail Design
+
+```mermaid
+flowchart LR
+subgraph ds[deviceShifu-LwM2M]
+    subgraph container
+        direction TB
+        lws[LwM2M Server]
```
[the remark this answers is not in the document]
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:31]: for now we don't support multiple device use same lwm2M server, the device's endpoint name should be same with it defined in edgedevice.
yes, for different lwM2M deviceshifu should expose different port

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+ap <-->|restful API| http
+ap <-->|restful API| http1
+lws <-->|lwM2M| d
+lws1 <-->|lwM2M| d1
+```
+
+The `deviceShifu-LwM2M` will represent each LwM2M object as an instruction that can be accessed through a RESTful API. The API will use the POST method for write operations and the GET method for read operations. The supported data formats for LwM2M v1.0 are TLV, JSON, Plain text, and Opaque.
+
+deviceShifu-LwM2M's LwM2M server will handle Register, Update, De-register, Read, Write, Observe and Notify.
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:55]: Register, Update, De-register, Read, Write, Observe and Notify.  there are some typo and duplication in this text.

correct one FYI: 
Register, Un-register, Read, Write, Observe and Notify.

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+For each deviceshifu lwM2M should expose a UDP port for LwM2M communication.
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:57]: make sure all words consist with a standard, should "LwM2M" be the most correct spelling?

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+deviceShifu-LwM2M support two kind of mode normal and observe mode.
+- normal mode: just like the other deviceshifu, when call the instruction, deviceShifu will send Read or Write Request to deviceShifu and return response
+- observe mode: this mode will set the min and max notify time, then device will notify data when data changed or timeout. deviceShifu will record data, and return the data store in the memory cache when call the instruction. and this kind of mode also support read and write operation.
+
+deviceShifu will host a LwM2M server and listen on udp 3683(LwM2M default port) and http server(deviceshifu) on 8080.
+
+the LwM2M server will handle register, update, de-register request from device and maintain the device info in the memory cache.
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:65]: where is read/write/observe/notify request here? should be consist with above statement

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+if the Object is in observe mode, LwM2M server also need to handle the notify request from device and update the data in the memory cache.
+
+when the deviceShifu received the instruction before the device register, it will return error message. 
+
+```mermaid
+sequenceDiagram
+    participant d as device
+    participant ds as DeviceShifu
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:74]: deviceShifu?

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    d ->> bs: Get Server Info
+    bs ->> d: Reply Server Info
+    end
+
+    d ->> ds: Register
+    ds ->> d: Created
+
+    alt observe mode
+    ds ->> d: Observe[Write the Notification Attributes Settings]
+    d ->> ds: success
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:88]: Success

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md (hunk 8 above)
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:55]: image
In the official lwM2M documentation, it is named this way.

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md (hunk 9 above)
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:57]: yes the `LwM2M` is correct

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md (hunk 10 above)
Comment (rhoninl) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:65]: this line is same with above removed

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+- Datagram Transport Layer Security (DTLS) support.
+- Resource observation or Notification.
+- [LwM2M protocol using v1.0.x version](https://www.openmobilealliance.org/release/LightweightM2M/V1_0-20170208-A/OMA-TS-LightweightM2M-V1_0-20170208-A.pdf).
+- LwM2M protocol under UDP.
+  
+### Design Non-Goal
+
+- Support LwM2M v1.1.x or later version.
+- Over TCP or other protocol.
+- Bootstrap Server.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:22]: These are features vs unsupported features, not design goals

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+          http1 -->|call interface| lws1
+  end
+end
+ap[application]
+d[device]
+d1[device1]
+
+
+ap <-->|restful API| http
+ap <-->|restful API| http1
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:48]: RESTful

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+- Support LwM2M v1.1.x or later version.
+- Over TCP or other protocol.
+- Bootstrap Server.
+
+## Detail Design
+
+```mermaid
+flowchart LR
+subgraph kubernetes[Kubernetes]
+  subgraph ds[deviceShifu-LwM2M]
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:29]: number the deviceShifu too

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+d1[device1]
+
+
+ap <-->|restful API| http
+ap <-->|restful API| http1
+lws <-->|LwM2M| d
+lws1 <-->|LwM2M| d1
+```
+
+The `deviceShifu-LwM2M` will represent each LwM2M object as an instruction that can be accessed through a RESTful API. The API will use the POST method for write operations and the GET method for read operations. The supported data formats for LwM2M v1.0 are TLV, JSON, Plain text, and Opaque.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:53]: put supported formats in a list

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+deviceShifu-LwM2M's LwM2M server will handle Register, Update, De-register, Read, Write, Observe and Notify. And maintain the device info in the memory cache.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:55]: list the functions

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+
+For each deviceshifu LwM2M should expose a UDP port for LwM2M communication.
+
+deviceShifu-LwM2M support two kind of mode normal and observe mode.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:59]: "two modes"

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+ap <-->|restful API| http1
+lws <-->|LwM2M| d
+lws1 <-->|LwM2M| d1
+```
+
+The `deviceShifu-LwM2M` will represent each LwM2M object as an instruction that can be accessed through a RESTful API. The API will use the POST method for write operations and the GET method for read operations. The supported data formats for LwM2M v1.0 are TLV, JSON, Plain text, and Opaque.
+
+deviceShifu-LwM2M's LwM2M server will handle Register, Update, De-register, Read, Write, Observe and Notify. And maintain the device info in the memory cache.
+
+For each deviceshifu LwM2M should expose a UDP port for LwM2M communication.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:57]: each deviceShifuLwM2M should...

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+
+deviceShifu-LwM2M support two kind of mode normal and observe mode.
+- normal mode: just like the other deviceshifu, when call the instruction, deviceShifu will send Read or Write Request to deviceShifu and return response
+- observe mode: this mode will set the min and max notify time, then device will notify data when data changed or timeout. deviceShifu will record data, and return the data store in the memory cache when call the instruction. and this kind of mode also support read and write operation.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:61]: link to there reference docs

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+deviceShifu will host a LwM2M server and listen on udp 3683(LwM2M default port) and http server(deviceshifu) on 8080.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:63]: UDP
HTTP
deviceShifu, same for others

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+
+if the Object is in observe mode, LwM2M server also need to handle the notify request from device and update the data in the memory cache.
+
+when the deviceShifu received the instruction before the device register, it will return error message.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:67]: capitalization, also fix grammar

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+        end
+    end
+```
+
+### Protocol Specification
+
+Define data structures and types in Go for configuring and managing LwM2M communication:
+
+```go
+type LwM2MSettings struct {
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:112]: LwM2MSetting

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+
+type CiperSuite string
+
+const (
+	CiperSuite_TLS_ECDHE_ECDSA_WITH_AES_128_CCM   CiperSuite = "TLS_ECDHE_ECDSA_WITH_AES_128_CCM"
+    // ...
+)
+```
+
+If bootstrap server address is not empty, deviceShifu will send register request to bootstrap server, and get the server info, then send register request to deviceShifu.
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:138]: remove if not supported

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+
+```go
+type LwM2MType string
+
+type Properties struct {
+    ObjectId            string      `json:"ObjectId"` // required example /3303/0
+    DataFormat          LwM2MType   `json:"dataFormat"` // optional TLV/JSON/PlainText/Opaque default plaintext
+    EnableObserve       bool        `json:"EnableObserve,omitempty"` // optional enable observe mode default false
+}
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:147]: need update

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+
+### Serving requests
+
+deviceShifu-LwM2M would take RESTful-style requests just as other deviceShifu do.
+
+LwM2M supports both `GET` and `PUT` requests.
+
+For read the data from device, the method signature looks like:
+```
+GET lwm2m.device.svc.cluster.local/{object1_name}
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:158]: instruction_name

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md (hunk 16 above)
Comment (tomqin93) [reply docs/design/deviceshifu/design-deviceShifu-lwm2m.md:53]: POST => PUT

**Review by tomqin93 — COMMENTED** [review]
@rhoninl added some comments, please take a look

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+
+```go
+type LwM2MType string
+
+type Properties struct {
+    ObjectId            string      `json:"ObjectId"` // required example /3303/0
+    DataType            LwM2MType   `json:"DataType"` // optional float, int, bool, string, default string
+    EnableObserve       bool        `json:"EnableObserve,omitempty"` // optional enable observe mode default false
+}
+```
```
Comment (tomqin93) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:157]: are these for `protocolPropertyList` in ConfigMap? if so add a section title to this file

**Review by tomqin93 — COMMENTED** [review]
@rhoninl added some comments, please take a look

**Review by tomqin93 — APPROVED** [review]
LGTM

File: docs/design/deviceshifu/design-deviceShifu-lwm2m.md [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+deviceShifu-LwM2M's LwM2M server will handle the following functions:
+- Register
+- Update
+- De-register
+- Read
+- Write
+- Observe
+- Notify
+
+Each deviceshifuLwM2M should expose a UDP port for LwM2M communication.
```
Comment (kris21he) [inline docs/design/deviceshifu/design-deviceShifu-lwm2m.md:69]: deviceshifu-LwM2M

**Comment by kris21he** [comment]
@rhoninl one more typo? please check, others LGTM

**Pull request merged**